interfaces/prompting/patterns/patterns.go

```diff
@@ -0,0 +1,184 @@
+// pattern and renders each one into a string which is then passed into the
+// given observe closure, along with the index of the variant.
+//
+// The given observe closure should perform some action with the rendered
+// variant, such as adding it to a data structure.
+func (p *PathPattern) RenderAllVariants(observe func(index int, variant string)) {
+	cleanThenObserve := func(i int, variant string) {
+		cleaned := cleanPattern(variant)
+		observe(i, cleaned)
+	}
+	renderAllVariants(p.renderTree, cleanThenObserve)
+}
+
+var (
+	duplicateSlashes    = regexp.MustCompile(`(^|[^\\])/+`)
+	charsDoublestar     = regexp.MustCompile(`([^/\\])\*\*+`)
+	doublestarChars     = regexp.MustCompile(`([^\\])\*\*+([^/])`)
+	duplicateDoublestar = regexp.MustCompile(`/\*\*(/\*\*)+`) // relies on charsDoublestar running first
+	starsAnyMaybeStars  = regexp.MustCompile(`([^\\])\*+(\?\**)+`)
+)
+
+func cleanPattern(pattern string) string {
+	pattern = duplicateSlashes.ReplaceAllString(pattern, `${1}/`)
+	pattern = charsDoublestar.ReplaceAllString(pattern, `${1}*`)
+	pattern = doublestarChars.ReplaceAllString(pattern, `${1}*${2}`)
+	pattern = duplicateDoublestar.ReplaceAllString(pattern, `/**`)
+	pattern = starsAnyMaybeStars.ReplaceAllStringFunc(pattern, func(s string) string {
+		deleteStars := func(r rune) rune {
+			if r == '*' {
+				return -1
+			}
+			return r
+		}
+		return strings.Map(deleteStars, s) + "*"
+	})
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-len("/*")]
+	}
+	return pattern
+}
+
+// PathPatternMatches returns true if the given pattern matches the given path.
+//
+// Paths to directories are received with trailing slashes, but we don't want
+// to require the user to include a trailing '/' if they want to match
+// directories (and end their pattern with `{,/}` if they want to match both
+// directories and non-directories). Thus, we want to ensure that patterns
+// without trailing slashes match paths with trailing slashes. However,
+// patterns with trailing slashes should not match paths without trailing
+// slashes.
+//
+// The doublestar package (v4.6.1) has special cases for patterns ending in
+// `/**` and `/**/`: `/foo/**`, and `/foo/**/` both match `/foo` and `/foo/`.
+// We want to override this behavior to make `/foo/**/` not match `/foo`.
+// We also want to override doublestar to make `/foo` match `/foo/`.
+func PathPatternMatches(pattern string, path string) (bool, error) {
+	// Check the usual doublestar match first, in case the pattern is malformed
+	// and causes an error, and return the error if so.
+	matched, err := doublestar.Match(pattern, path)
+	if err != nil {
+		return false, err
+	}
+	// No matter if doublestar matched, return false if pattern ends in '/' but
+	// path is not a directory.
+	if strings.HasSuffix(pattern, "/") && !strings.HasSuffix(path, "/") {
+		return false, nil
+	}
```

Comment:
if you're discarding match result, then could this be done earlier?

Reply:
The first match result happens first to check that everything is valid, and throw an error if not. Then (assuming there was not an error), the match result is used to see if the pattern as-is matches the path. If there's no match, we then check if the pattern would match if it ended in a `'/'`. So the match result isn't really discarded, just some reorganization to handle the error first.

But the pattern should be known to be valid, since we rendered it, so perhaps this check could happen before the first match. And this would save the computation time of `doublestar.Match` if we know it doesn't match (by our standards).